File: ClassesDemo/src/Knapsack.java
import java.util.ArrayList;

public class Knapsack
{
    private ArrayList<Task> tasks;
    private int capacity;

    public Knapsack(ArrayList<Task> tasks, int capacity)
    {
        this.tasks = tasks;
        this.capacity = capacity;
    }

    public Solution solve()
    {
        int numOfTasks = tasks.size();
        int[][]mat = new int[numOfTasks + 1][capacity + 1];

        for (int i = 0; i <= capacity; i++)
            mat[0][i] = 0;
        
        for (int i = 1; i <= numOfTasks; i++)
        {
            for (int j = 0; j <= capacity; j++)
            {
                if (tasks.get(i - 1).getHours() > j)
                    mat[i][j] = mat[i - 1][j];
                else
                    mat[i][j] = (int) Math.max(
                        mat[i - 1][j],
                        mat[i - 1][j - (int) tasks.get(i - 1).getHours()] + (int) tasks.get(i - 1).getValue());
            }
        }

        int result = mat[numOfTasks][capacity];
        int w = capacity;
        ArrayList<Task> tasksSolution = new ArrayList<>();

        for (int i = numOfTasks; i > 0 && result > 0; i--)
        {
            if (result != mat[i - 1][w])
            {
                tasksSolution.add(tasks.get(i - 1));
                result -= (int) tasks.get(i - 1).getValue();
                w -= (int) tasks.get(i - 1).getHours();
            }
        }

        return new Solution(tasksSolution, mat[numOfTasks][capacity]);
    }

    // ===========================================================================

    public void display()
    {
        if (tasks != null && !tasks.isEmpty())
        {
            System.out.println("Knapsack problem:");
            System.out.println("Capacity = " + capacity);
            System.out.println("Tasks: ");

            for (Task task : tasks)
                System.out.println("- " + task);
        }
    }

    // ===========================================================================
}

File: src/services/knapsack.ts
import { Task } from './task'
import { Solution } from './solution'

interface Knapsack {
    tasks: Task[]
    capacity: number
}

function solve(knapsack: Knapsack): Solution {
    const numOfTasks = knapsack.tasks.length
    const capacity = knapsack.capacity
    console.log('capacity = ' + capacity)

    let matrix = new Array(numOfTasks + 1)

    for (let i = 0; i < matrix.length; i++)
        matrix[i] = new Array(capacity + 1).fill(0)

    for (let i = 1; i <= numOfTasks; i++) {
        for (let j = 0; j <= capacity; j++) {
            if (knapsack.tasks[i - 1].hours > j) {
                matrix[i][j] = matrix[i - 1][j]
            } else {
                matrix[i][j] = Math.max(
                    matrix[i - 1][j],
                    matrix[i - 1][j - Math.round(knapsack.tasks[i - 1].hours)] +
                        knapsack.tasks[i - 1].value
                )
            }
        }
    }

    let result = matrix[numOfTasks][capacity]
    let w = capacity

    let tasksSolution: Task[] = new Array()
    let numOfHours: number = 0

    for (let i = numOfTasks; i > 0 && result > 0; i--) {
        if (matrix[i - 1][w] !== result) {
            tasksSolution.push(knapsack.tasks[i - 1])
            numOfHours += knapsack.tasks[i - 1].hours

            result -= knapsack.tasks[i - 1].value
            w -= knapsack.tasks[i - 1].hours
        }
    }

    const solution: Solution = {
        value: matrix[numOfTasks][capacity],
        hours: numOfHours,
        tasks: tasksSolution,
    }

    return solution
}

export { Knapsack, solve }
